File: SamsidParty.Subsonic.Proxy.AppleMusic/Types/Song.cs
using Newtonsoft.Json;
using SamsidParty.Subsonic.Common;
using SamsidParty.Subsonic.Common.Types;

namespace SamsidParty.Subsonic.Proxy.AppleMusic.Types
{
    public class Song : Resource
    {
        [JsonProperty("type")]
        public new string Type { get; set; } = "songs";

        [JsonProperty("attributes")]
        public SongAttributes Attributes { get; set; }

        [JsonProperty("relationships")]
        public SongRelationships Relationships { get; set; }

        public Child ToSubsonicType()
        {
            return new Child()
            {
                Id = "applemusic\n" + Id + "\n" + Attributes.PlayParams.Id,
                IsDir = false,
                IsVideo = false,
                Type = GenericMediaType.Music,
                Title = Attributes.Name,
                Album = Attributes.AlbumName,
                Artist = Attributes.ArtistName,
                DisplayAlbumArtist = Attributes.ArtistName,
                DiscNumber = Attributes.DiscNumber,
                Track = Attributes.TrackNumber,
                Year = int.Parse(Attributes.ReleaseDate.Substring(0, 4)),
                ContentType = "audio/m4a",
                Suffix = "m4a",
                Duration = Attributes.DurationInMillis / 1000,
                CoverArt = Attributes.Artwork.Url
            };
        }
    }

    public class SongAttributes
    {
        [JsonProperty("albumName")]
        public string AlbumName { get; set; }

        [JsonProperty("artistName")]
        public string ArtistName { get; set; }

        [JsonProperty("artwork")]
        public Artwork Artwork { get; set; }

        [JsonProperty("composerName")]
        public string ComposerName { get; set; }

        [JsonProperty("contentRating")]
        public string ContentRating { get; set; }

        [JsonProperty("discNumber")]
        public int DiscNumber { get; set; }

        [JsonProperty("durationInMillis")]
        public int DurationInMillis { get; set; }

        [JsonProperty("editorialNotes")]
        public EditorialNotes EditorialNotes { get; set; }

        [JsonProperty("genreNames")]
        public System.Collections.Generic.List<string> GenreNames { get; set; }

        [JsonProperty("hasLyrics")]
        public bool HasLyrics { get; set; }

        [JsonProperty("isrc")]
        public string Isrc { get; set; }

        [JsonProperty("movementCount")]
        public int? MovementCount { get; set; }

        [JsonProperty("movementName")]
        public string MovementName { get; set; }

        [JsonProperty("movementNumber")]
        public string MovementNumber { get; set; }

        [JsonProperty("name")]
        public string Name { get; set; }

        [JsonProperty("playParams")]
        public PlayParameters PlayParams { get; set; }

        [JsonProperty("previews")]
        public System.Collections.Generic.List<Preview> Previews { get; set; }

        [JsonProperty("releaseDate")]
        public string ReleaseDate { get; set; }

        [JsonProperty("trackNumber")]
        public int TrackNumber { get; set; }

        [JsonProperty("url")]
        public string Url { get; set; }

        [JsonProperty("workName")]
        public string WorkName { get; set; }
    }
}

File: FocalSonic/src-vite/src/types/applemusic/song.ts
import removeUndefined from "@/utils/removeUndefined";
import { Resource } from "i18next";
import { AppleMusicAlbum } from "./albums";
import { AppleMusicArtist } from "./artist";
import { AppleMusicArtwork, AppleMusicEditorialNotes, AppleMusicGenre, AppleMusicPlayParams, AppleMusicRelationship } from "./common";

export interface AppleMusicLyricsResponse {
    data: AppleMusicLyrics[];
}

export interface AppleMusicLyrics {
    attributes: AppleMusicLyricsAttributes;
}

export interface AppleMusicLyricsAttributes {
    ttml?: string;
}

export interface AppleMusicSong extends Resource {
    type: "songs";
    // https://developer.apple.com/documentation/applemusicapi/song/attributes
    attributes?: {
        albumName: string;
        artistName: string;
        artwork?: AppleMusicArtwork | undefined;
        composerName?: string | undefined;
        contentRating?: string | undefined;
        discNumber: number;
        durationInMillis: number;
        editorialNotes?: AppleMusicEditorialNotes | undefined;
        genreNames: string[];
        hasLyrics: boolean;
        isrc: string;
        movementCount?: number | undefined;
        movementName?: string | undefined;
        movementNumber?: string | undefined;
        name: string;
        playParams?: AppleMusicPlayParams | undefined;
        previews: AppleMusicPreview[];
        releaseDate: string;
        trackNumber: number;
        inFavorites: boolean | undefined;
        url: string;
        workName?: string | undefined;
    } | undefined;
    relationships?: AppleMusicSongRelationships | undefined;
}

// https://developer.apple.com/documentation/applemusicapi/song/relationships
export interface AppleMusicSongRelationships {
    albums: AppleMusicRelationship<AppleMusicAlbum>;
    artists: AppleMusicRelationship<AppleMusicArtist>;
    genres?: AppleMusicRelationship<AppleMusicGenre> | undefined;
    station?: { data: AppleMusicStation } | undefined;
}

export function convertAppleMusicSongToSubsonic(song: AppleMusicSong, parent: any | undefined): Song {
    if (!song) { return; }

    return removeUndefined({
        isDir: false,
        id: song.attributes?.playParams?.catalogId || song.attributes?.playParams?.id || song.id,
        parent: song.attributes?.playParams?.catalogId || song.attributes?.playParams?.id || song.id,
        albumId: song.attributes?.playParams?.catalogId || song.attributes?.playParams?.id || song.id,
        title: song.attributes?.name || "Unknown",
        name: song.attributes?.name || "Unknown",
        artist: song.attributes?.artistName || "Unknown",
        artistId: "authorof:" + (song.attributes?.playParams?.catalogId || song.attributes?.playParams?.id || song.id),
        album: song.attributes?.albumName || "Unknown",
        duration: Math.ceil((song.attributes?.durationInMillis || 0) / 1000),
        suffix: "m4a",
        coverArt: song.attributes?.artwork?.url || "",
        starred: song.attributes?.inFavorites === true ? new Date().toISOString() : undefined,
        explicitStatus: song.attributes?.contentRating || "clean",
        appleMusic: {
            data: song,
            libraryID: song.id,
            parent: parent
        }
    });
}
